staging/src/k8s.io/apimachinery/pkg/api/meta/help.go

```diff
@@ -202,8 +264,56 @@ func ExtractList(obj runtime.Object) ([]runtime.Object, error) {
 	return list, nil
 }
 
-// objectSliceType is the type of a slice of Objects
-var objectSliceType = reflect.TypeOf([]runtime.Object{})
+// ExtractListWithAlloc returns obj's Items element as an array of runtime.Objects.
+// the difference from ExtractList is that an object of the same type is regenerated and assigned a value,
+// so that the returned Object list and obj have independent life cycles.
+// Returns an error if obj is not a List type (does not have an Items member).
+func ExtractListWithAlloc(obj runtime.Object) ([]runtime.Object, error) {
+	itemsPtr, err := GetItemsPtr(obj)
+	if err != nil {
+		return nil, err
+	}
+	items, err := conversion.EnforcePtr(itemsPtr)
+	if err != nil {
+		return nil, err
+	}
+	list := make([]runtime.Object, items.Len())
+	for i := range list {
+		raw := items.Index(i)
+		switch {
+		case raw.Type() == rawExtensionObjectType:
+			item := raw.Interface().(runtime.RawExtension)
+			switch {
+			case item.Object != nil:
+				list[i] = item.Object
+			case item.Raw != nil:
+				// TODO: Set ContentEncoding and ContentType correctly.
+				list[i] = &runtime.Unknown{Raw: item.Raw}
+			default:
+				list[i] = nil
+			}
+		case raw.Type().Implements(objectType):
+			list[i] = raw.Interface().(runtime.Object)
+		default:
+			// shallow copy to avoid retaining a reference to the original list item
+			new := reflect.New(raw.Type())
```

Comment:
Does this consume that much more memory? would people ever want to call the old version?

Also please don't use `new` as a name, it shadows the built-in `new()` function which is confusing.

Reply:
thanks for the review suggestion, it has been modified. PTAL...

Reply:
> Does this consume that much more memory?

No, https://github.com/kubernetes/kubernetes/pull/113362#issuecomment-1297108169 The discussion here can provide some references.

Memory allocation does happen here. But not as much as `Deepcopy`. Because what happens here is a shallow copy.

With the complex type of Pod, its fields have rich types:

- map type: such as Annotations or Labels, ...
- sliceType: spec.Containers, spec.InitContainers, status.Conditions, ...

The above types still use the allocated memory in the items element.

But simple types (eg: string, int64): Name, Namespace, Generation... will be reassigned.

I think the price is acceptable. (Even without this change, the code of the master branch will still have this memory allocation)

> would people ever want to call the old version?

Yes, so we provide two functions, namely `ExtractListWithAlloc` and `ExtractList` for users to make choices.

Reply:
>> would people ever want to call the old version?

> Yes

Why?

Reply:
The difference between ExtractList (optimized) and ExtractListWithAlloc is:

ExtractListWithAlloc will use a `shallow copy` to generate a new Object. And ExtractList directly refers to the elements in Items.


Take `PodList` as an example: 

Assuming that in the logic of replicaset-controller's processing of Pods, our `ListAndWatcher` will extract the PodList, and then put pods[0], pods[1], pods[2] into the `Index`. The advantage of using ExtractWithAlloc at this time is that pods[0], pods[1], pods[2] will have an independent life cycle. When pods[0]' (the next state of pod[0]) comes, pod [0] can be GC.


For example, ExtractList used in DeleteCollection.

https://github.com/kubernetes/kubernetes/blob/v1.26.0-rc.1/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1137-L1148


Since this is a synchronous http request. The life cycle of podList has been limited to this request-response. Then we use ExtractList, pod[0], pod[1], pod[2] to directly refer to the elements in podList, which reduces memory allocation compared to ExtractListWithAlloc.

Summarize:

1. If the life cycle of podList and its elements has been limited to a synchronization context, then it is appropriate to use ExtractList
2. If the elements after the podList are extracted will be processed asynchronously, then it is appropriate to use ExtractWithAlloc.

Reply:
I think @lavalamp is asking if there's a reason not to switch the existing ExtractList to the new behavior unconditionally, since we've offset the increased allocation cost with the improvements to reflect usage.

A compelling reason not to change the existing functions is that callers can currently modify items returned by the functions and have those modifications change the items in the original list as well. That's an externally visible behavior that _requires_ referencing the memory locations in the original list.

Having new functions:
1. lets us switch to use those only in places where we know we're not expecting to modify items in the original list (like informers)
2. actually improves performance for places where we extract list items and don't retain them long-term (since we get the benefit of the reflect improvements without the hit of the new allocs)

Reply:
Thanks for your explanation.

Reply:
> A compelling reason not to change the existing functions is that callers can currently modify items returned by the functions and have those modifications change the items in the original list as well.

Well... the new behavior of shallow copying is halfway like that, it's still not suitable if you *don't* want that behavior.

But I am assuming that this behavior is unwanted and people have to work around it, so doing a shallow copy unconditionally doesn't make it worse. Can you find a place where people are deliberately modifying a list in-place via the result of these functions?

Reply:
@liggitt @lavalamp 
If it is convenient, please tell me, what should I do to promote this PR?